akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala

```diff
@@ -38,62 +42,75 @@ private[http] object OutgoingConnectionBlueprint {
-      .map(RequestRenderingContext(_, hostHeader))
-      .via(Flow[RequestRenderingContext].flatMapConcat(requestRendererFactory.renderToSource).named("renderer"))
+      val bypassFanout = b.add(Broadcast[RequestRenderingContext](2, eagerCancel = true))
 
-    val methodBypass = Flow[HttpRequest].map(_.method)
+      val terminationMerge = b.add(TerminationMerge)
 
-    import ParserOutput._
-    val responsePrep = Flow[List[ResponseOutput]]
-      .mapConcat(conforms)
-      .via(new PrepareResponse(parserSettings))
+      val requestRendering: Flow[RequestRenderingContext, ByteString, NotUsed] = {
+        val requestRendererFactory = new HttpRequestRendererFactory(userAgentHeader, requestHeaderSizeHint, log)
+        Flow[RequestRenderingContext].flatMapConcat(requestRendererFactory.renderToSource).named("renderer")
+      }
 
-    val core = BidiFlow.fromGraph(GraphDSL.create() { implicit b ⇒
-      import GraphDSL.Implicits._
-      val methodBypassFanout = b.add(Broadcast[HttpRequest](2, eagerCancel = true))
-      val responseParsingMerge = b.add(new ResponseParsingMerge(rootParser))
+      val bypass = Flow[RequestRenderingContext] map { ctx ⇒
+        HttpResponseParser.ResponseContext(ctx.request.method, ctx.sendEntityTrigger.map(_.asInstanceOf[Promise[Unit]]))
+      }
+
+      val responseParsingMerge = b.add {
+        // the initial header parser we initially use for every connection,
+        // will not be mutated, all "shared copy" parsers copy on first-write into the header cache
+        val rootParser = new HttpResponseParser(parserSettings, HttpHeaderParser(parserSettings) { info ⇒
+          if (parserSettings.illegalHeaderWarnings)
+            logParsingError(info withSummaryPrepended "Illegal response header", log, parserSettings.errorLoggingVerbosity)
+        })
+        new ResponseParsingMerge(rootParser)
+      }
+
+      val responsePrep = Flow[List[ParserOutput.ResponseOutput]]
+        .mapConcat(conforms)
+        .via(new PrepareResponse(parserSettings))
 
       val terminationFanout = b.add(Broadcast[HttpResponse](2))
-      val terminationMerge = b.add(TerminationMerge)
 
       val logger = b.add(MapError[ByteString] { case t ⇒ log.error(t, "Outgoing request stream error"); t }.named("errorLogger"))
       val wrapTls = b.add(Flow[ByteString].map(SendBytes))
-      terminationMerge.out ~> requestRendering ~> logger ~> wrapTls
 
       val collectSessionBytes = b.add(Flow[SslTlsInbound].collect { case s: SessionBytes ⇒ s })
-      collectSessionBytes ~> responseParsingMerge.in0
 
-      methodBypassFanout.out(0) ~> terminationMerge.in0
+      renderingContextCreation.out ~> bypassFanout.in
+      bypassFanout.out(0) ~> terminationMerge.in0
+      terminationMerge.out ~> requestRendering ~> logger ~> wrapTls
 
-      methodBypassFanout.out(1) ~> methodBypass ~> responseParsingMerge.in1
+      bypassFanout.out(1) ~> bypass ~> responseParsingMerge.in1
```

Comment:
Oh I see, it took me a bit to see why the dance of Promise.future and then casting back to Promise.
So the bypass branch may complete that promise, and the other side may take action on it...